File: plugins/collector/event/apps/database/sqlserver.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
)

type Sqlserver struct {
	version string
	reg     *regexp.Regexp
}

func (Sqlserver) Name() string { return "sqlserver" }

func (Sqlserver) Type() string { return "database" }

func (e Sqlserver) Version() string { return e.version }

func (Sqlserver) Match(p *process.Process) bool { return p.Name == "sqlserver" }

func (e *Sqlserver) Run(p *process.Process) (mapping map[string]string, err error) {
	if e.reg == nil {
		e.reg = regexp.MustCompile(`(\d+\.){3}\d+`)
	}
	result, err := apps.Execute(p, "-v")
	if err != nil {
		return nil, err
	}
	str := e.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	return
}

func init() {
	apps.Regist(&Sqlserver{})
}

File: plugins/collector/src/event/apps/database/sqlserver.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Sqlserver {
    version: String,
}

impl Sqlserver {
    pub fn new() -> Self { Sqlserver { version: String::new() } }
}

impl IApp for Sqlserver {
    fn name(&self)     -> &'static str { "sqlserver" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "sqlserver" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-v"])?;
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
